### ISSUE TITLE
Option to set resolution to scale

### ISSUE THREAD

**Issue by motoleon** [Blinue/Magpie#278]
### Expected behavior 预期的功能

An option to set the max resolution to scale. Today it scales to max monitor resolution (desktop). But if you put an option to set "resolution to scale", it will be better. I have nvidia 1060 and a 3k monitor, but it cant render well to 3k. I want to put 1080p to scale but no option, without change desktop resolution. But i want to leave desktop resolution to 3k and play games to 1080p. Nvidia NIS, DSLL and AMD FSR not have this option. It all scales to max monitor resolution. It is an error, i think. Less power gpus cant render at max resolution monitor, sometimes...

**Comment by hooke007** [comment]
>  I want to put 1080p to scale but no option, without change desktop resolution. 

You have to manually change your desktop resolution via system settings / graphic driver. The scaled window is still a "borderless window", not an exclusive program.
Set the target to any res lower than desktop res, and magpie would add black bars.

**Comment by motoleon** [comment]
I know that. But i want my desktop in 3k and play games at max 2k or same, because my gpu card not runs ok with 3k. So an option with max scale-to resolution (in my case, 2k) will resolve my problem (and other users). My desktop runs ok at 3k resolution (monitor native), but not games. it is very difficult so implement?

**Comment by hooke007** [comment]
I'm not sure about your actual requirements.

> Set the target to any res lower than desktop res, and magpie would add black bars.

It didn't work for you?

**Comment by motoleon** [comment]
No. I dont want to change my desktop resolution. It is 3k and it is ok.
I want magpie to scale to xxx resolution, for example 2k or 1080p(full hd), not 3k. So, magpie may be able to change visible resolution on the fly.

**Comment by hooke007** [comment]
> Set the target to any res lower than desktop res, and magpie would add black bars.

I do not make you change desktop res here.
Just change the related parameter such as `"scale":[-1,-1]` in ScaleModels.json

**Comment by motoleon** [comment]
You dont understand my request.

**Comment by hooke007** [comment]
CC @Blinue

**Comment by bloc97** [comment]
> No. I dont want to change my desktop resolution. It is 3k and it is ok. I want magpie to scale to xxx resolution, for example 2k or 1080p(full hd), not 3k. So, magpie may be able to change visible resolution on the fly, for example 2k.

So if I understand correctly you have let's say a 720p game and 1440p monitor, you want to scale 720p -> 1080p using FSR and then display it without black bars on the 1440p screen?
You can set the target resolution of FSR using something like `"scale":[-0.8,-0.8]` then use bilinear or lanczos upscaling to 1440p.
The result will be 720p Game -> FSR -> 1080p -> Bilinear -> 1440p.

**Comment by motoleon** [comment]
bloc97, i have 720p game, and 3k monitor. I want to scale 720p-> 1080p using FSR and display it, at 1080p resolution, not more.
My gtx 1060 cant run games at native 3k, nor upscale with fsr, nvidia nis, none. It cant. It is a 1080p gpu, but my monitor is 34" ultrawide 3k.

**Comment by bloc97** [comment]
Well to display a 1080p image on a 1440p screen without black bars you have to upscale somewhere. Either it's Magpie that does the upscaling, or it's the GPU Encoder (some drivers have a setting for this called GPU Scaling or something similar), or it's the screen (when the GPU's output is lower than screen resolution). 

Magpie does not currently have access to the drivers and does not have a exclusive fullscreen mode. I don't know if it's feasible but @Blinue might know more.

**Comment by Blinue** [comment]
Without the help of certain technologies (such as GPU scaling) you'll have to set the screen resolution. Usually when a game is 'fullscreen', it means that it has the same resolution as the monitor.

So, the solution is:

Set the screen resolution to 1080p -> Run the game in full screen (1080p) -> Set the screen resolution back to 3k

You want Magpie to perform the first and third steps automatically. Am I right?

**Comment by motoleon** [comment]
I have desktop in 3k. I start the game in 720p windowed. I want Magpie when is activated,  to set screen resolution to 1080p (configured parameter), and it scales the game to 1080p.

**Comment by 0x4E69676874466F78** [comment]
@motoleon it is strange that your card can give out a desktop in 3k but cannot scale through Magpie in 3k.

**Comment by motoleon** [comment]
> @motoleon it is strange that your card can give out a desktop in 3k but cannot scale through Magpie in 3k.

No, no, my gpu CANT render games at 3k, because LOW FPS. This is the problem. With or without Magpie.

**Comment by motoleon** [comment]
Nvidia 1060 3Gb is not made to run games at 3k-4k resolution. It is a 1080p gpu max. So i want to limit output resolution. It is a common problem today with out of stock gpus, we use old gpus to run games...in newer hardware (monitors)

**Comment by 0x4E69676874466F78** [comment]
Well, I'm talking about Mapgie. Is scaling to 3k really that expensive for 1060?

**Comment by motoleon** [comment]
> Well, I'm talking about Mapgie. Is scaling to 3k really that expensive for 1060?

yes. And without Magpie too.

**Comment by 0x4E69676874466F78** [comment]
Without magpie, this is understandable (there the game is rendered in high resolution and consumes noticeably more video memory). But Mapgie is just post-processing, it should not consume too much video memory.

**Comment by 0x4E69676874466F78** [comment]
Although I checked yes there is a problem, Magpie consumes quite a lot of video memory and it got a leak :(

**Comment by Blinue** [comment]
> Although I checked yes there is a problem, Magpie consumes quite a lot of video memory and it got a leak :(

I'm sure there are no leaks here, because the D3D debug layer reports these errors.

As the screen resolution increases, the video memory occupied by Magpie grows exponentially.

**Comment by 0x4E69676874466F78** [comment]
@Blinue 
> I'm sure there are no leaks here, because the D3D debug layer reports these errors.
Yes, sorry, I was confused by a bug in process hacker (wrong video memory counter).

**Comment by Blinue** [comment]
> I have desktop in 3k. I start the game in 720p windowed. I want Magpie when is activated, to set screen resolution to 1080p (configured parameter), and it scales the game to 1080p. When i stops the game or Magpie, Magpie goes back screen resolution to 3k (desktop resolution i have). The final scale (1080p->3k) is made by monitor, the output resolution by gpu is 1080p.

I don't think this is a common need, in fact this feature has nothing to do with Magpie's core functionality. You might be looking for other software, like [Reso](http://www.bcheck.net/apps/reso.htm).

**Comment by motoleon** [comment]
The mayority gpus in Steam are 1060. Today monitors are 2k-4k. This gpu is 1080p mid-low fps performance...it is a common need.
I think my petition is easy to implement. A solution to use low-end gpus on high-end monitors, limiting max resolution on those monitors..i think. Magpie will be a great utility until gpu market get normal prices.

**Comment by bloc97** [comment]
I just find it hard to believe that a 1060 would have difficulty upscaling an image from 1080p to 1440p using shaders. We're talking about a modern architecture that supports advanced unified shaders and has 4TFLOPS of theoretical performance. It's not like you're using a GT710... There must be something wrong somewhere as 1440p is only 1.7x the number of pixels of 1080p...

**Comment by bloc97** [comment]
I just ran a quick test and magpie consumes ~90MB of VRAM for 1080p->4K with Lanczos and ~141MB of VRAM with Anime4K 3D, this would be reduced to ~40MB and ~60MB for 1080p->1440p respectively. If TFLOPS and VRAM shouldn't be an issue, there's definitively something wrong.

**Comment by motoleon** [comment]
No, no. 1060 have difficulty to RENDER a game at 3K (3***X1440). No upscaling an image. I have not said that. Sorry my english, i am spanish.

**Comment by bloc97** [comment]
You said that rendering the game at 720p and then upscaling it using magpie to 1440p causes lag? If that's the case it should not happen. Your GPU should be perfectly fine.

**Comment by motoleon** [comment]
> You said that rendering the game at 720p and then upscaling it using magpie to 1440p causes lag? If that's the case it should not happen. Your GPU should be perfectly fine.

I not said that.

**Comment by Predator4578** [comment]
but u said u starting the game in 720p windowed? ;-P

anyways i think i have the same issue here, and its the issue that Magpie cant apply FSR proper BIGGER THAN 16/9 resolutions? with my ultrawide setup i cant use magpie because nothing happens when i upscale from Resolutions other than 16/9 aspect Ratio. i described more details here: https://github.com/Blinue/Magpie/issues/75 
your answer were: "Magpie currently has no plans to support widescreen."

Might add the possibility to Use also Widescreen Monitors in Future? I would donate to this Project for that even ;-)
btw there exist games with FSR option INSIDE the game and they can apply it to my Resolution of 7850x1440p... 
so it is definitively a Magpie issue here

TBH what Motoleon told you is the Truth for latest Triple A Titles, Many GPUs struggle at Ultrawide Resolutions because the GPU Core is already running 99% Usage on FullHD 1080p or WQHD 2560p for those Titles in Running only at about 60FPS
It is mostly feeling same shit in those games, not sure even if the developers just spam assets inside the level upon they can still achieve 60FPS with the Latest gpu, then the Game level is designed acceptable for them? mostly poorly optimized etc in these days too but thats another story., 

Now in my Case i have 3 x 144Hz Monitors and OFC i am looking for a Way to achieve 144 FPS 
or at least over 100 FPS in a game. Usually this can be achieved while using the Ingame Scalers and or use Dynamic Resolution of 60-80% of original only, sadly Nvidia only lets us "Superresolution" currently with its Driver meaning its the opposite of what i need, rendering the game in Factor 1 up to 2.5 x the Original ingame Resolution wich massive causes Stutter in these wide resolutions... Sadly not every game lets u choose if the ingame Rendered Resolution will be LOWER than the actual Desctop Resolution, for example Battlefield (one,V,2042) they all own this feature ingame or Free to play Warframe too
the Latest BF2042 can even Apply DLSS on ultrawide Resolutions, but its only a small number of games support that now

"Without magpie, this is understandable (there the game is rendered in high resolution and consumes noticeably more video memory). But Mapgie is just post-processing, it should not consume too much video memory." 
It is not that the Card has not enough VRAM or anyone is messing up and saying magpie takes too much VRAM, its simple because the GPU Core is maxed out in those titles

There is now the Story where FSR becomes interesting because u can Play a Game in A Lower Resolution and let it looking Good
while still perform Good aswell, it might not look as good as Native but it will look way better than the Lower Resolution and thats a "Accepted" Parameter for Many Ultra Wide Monitor Users, now the Sad Part that Magpie is not the Right Programm 
NOW for that, but i am sure Technically u could easily add this Feature, so i not get why u have so less interest in this too...

there would be Tons of thankful People might donating afterwards....

**Comment by motoleon** [comment]
Yes, i have ultra wide monitor, 3k. The problem is 1060 dies with that resolution, so i want to use 1-2k.

**Comment by onyxtrooper** [comment]
I've read the whole thread, but I still can't figure how to set a target resolution (regardless of the desktop resolution).

I have a 1440p display and I use 2x DSR on every game, which is 3620x2036 in resolution. If I change my desktop resolution to this, Magpie will upscale the games without any issues but changing desktop resolution every time I open a game is a bit frustrating.

I tried editing ScaleModels.json file and changed "scale":[-1,-1] to "scale":[-2,-2], but that resulted in an error. I don't have any programming knowledge, so I don't know how to progress further.

Little help please?

**Comment by hooke007** [comment]
@onyxtrooper Theoretically, it is equivalents that set `"scale":[2,2]` (the number is up to you, make it larger than your native res) first and then append a downscaling method to recover `"scale":[-1,-1]`

**Comment by onyxtrooper** [comment]
I'm using jsoneditoronline.org to edit the file. Here is a pic:

https://imgur.com/a/WDSMuvJ

Am I doing it right? I can't find where to enter -1,-1 though. There is only scale option under FSR

**Comment by hooke007** [comment]
a simple example
```json
  {"name":"3x_FSR_EASU--RCAS--Bicubic_Lite","effects":[
    {"effect":"FSR_EASU","scale":[3,3]},
    {"effect":"FSR_RCAS","sharpness":0.87},
    {"effect":"Bicubic_Lite","scale":[-1,-1],"paramB":0,"paramC":0.5},
  ]},
```

**Comment by onyxtrooper** [comment]
Should I copy paste this to the location "name": "FSR", ? Apologies for asking many questions, I'm really a noob at this

**Comment by hooke007** [comment]
copy the full content into that file. Put it between two "groups". i.e. L8-L9

https://github.com/Blinue/Magpie/blob/e552e60621e6acbf9d87fec1770e797dcf74f546/Magpie/Resources/BuiltInScaleModels.json#L1-L20

**Comment by onyxtrooper** [comment]
Thank you very much, I was able to select the new option "3x_FSR_EASU--RCAS--Bicubic_Lite" in the menu when I started Magpie.

There is something weird though, it caps the FPS at 20. I thought that my PC wasn't strong enough but I remembered I got like 45-50 FPS in 4k DSR. So I checked the FPS counter again, it never changes from 20, not even 19 or 21, so it must be a lock. I also realized G-SYNC icon disappeared when I started scaling. Could it be related to that?

**Comment by motoleon** [comment]
> I've read the whole thread, but I still can't figure how to set a target resolution (regardless of the desktop resolution).
> 
> I have a 1440p display and I use 2x DSR on every game, which is 3620x2036 in resolution. If I change my desktop resolution to this, Magpie will upscale the games without any issues but changing desktop resolution every time I open a game is a bit frustrating.
> 
> I tried editing ScaleModels.json file and changed "scale":[-1,-1] to "scale":[-2,-2], but that resulted in an error. I don't have any programming knowledge, so I don't know how to progress further.
> 
> Little help please?

With new parameter "Target resolution", magpie would put desktop resolution to that, scale up game and play.  That is.

**Comment by hooke007** [comment]
> I thought that my PC wasn't strong enough but I remembered I got like 45-50 FPS in 4k DSR.

I don't know your game's window res and it is just an example. In 1st step, 3x 1080p would be much larger than so-called 4k

**Comment by onyxtrooper** [comment]
> > I've read the whole thread, but I still can't figure how to set a target resolution (regardless of the desktop resolution).
> > I have a 1440p display and I use 2x DSR on every game, which is 3620x2036 in resolution. If I change my desktop resolution to this, Magpie will upscale the games without any issues but changing desktop resolution every time I open a game is a bit frustrating.
> > I tried editing ScaleModels.json file and changed "scale":[-1,-1] to "scale":[-2,-2], but that resulted in an error. I don't have any programming knowledge, so I don't know how to progress further.
> > Little help please?
> 
> With new parameter "Target resolution", magpie would put desktop resolution to that, scale up game and play. That is.

There is an option for that? I checked ScaleModels.json but couldn't find it. Can you tell me how?

**Comment by motoleon** [comment]
No, this option does not exist. It is my query for new enhancement

**Comment by bloc97** [comment]
> Tried changing in game resolution, it didn't have any effect. Correct me if I'm wrong, Magpie upscales based on the chosen resolution in game, right? If that's the case then my framerate is locked. It's stuck on 20 and it never changes under any circumstances.

Magpie has no access to in game settings, it only knows the size of the window. The default FSR setting upscales the window size to screen/desktop size. For example a windowed game with 1280x720 as window size will be upscaled to your screen size (eg. 1440p). You can edit ScaleModels.json so that it does not upscale to screen size and leave in black bars. If you don't want black bars you can use a ultra fast upscaling algorithm like bicubic after applying FSR.

**Comment by kanjieater** [comment]
> You can set the window resolution with an autohotkey macro.
> 
> See here: [autohotkey.com/docs/commands/WinMove.htm](https://www.autohotkey.com/docs/commands/WinMove.htm)
> 
> I have 3440x1440 but I'm trying to downscale an application while still having it full screen. So my Winmove is something like this:
> 
> `WinMove, (Window title), 0, 0, 1720, 720`
> 
> Put the macro file in the same folder as Magpie. Open game, run the macro, run Magpie.
> 
> > **[WinMove - Syntax & Usage | AutoHotkey](https://www.autohotkey.com/docs/commands/WinMove.htm)**The WinMove command changes the position and/or size of the specified window.

Are you saying your application was 3440x1440, and with the WinMove command you were able to resize it? What settings do you have in Magpie? Is your desired downscaled resolution 1720 x 720?

**Comment by regs01** [comment]
> No, no, my gpu CANT render games at 3k, because LOW FPS, and low vram. This is the problem. With or without Magpie.

And it won't. Your is still rendered in 720p. Magpie just stretching your your window to desktop resolution, using upscale technics. It does not render anything. Your game will still be rendered in 720p.

**Comment by Devilmonk7402** [comment]
So reading through this whole thread ( trying to understand the same thing as
